---
title: "Fluidigm SNP Array Report"
date:  "Report generated: `r Sys.Date()`"
params: 
   fluidigm_file: NULL
   somalier_file: NULL
   json_file: NULL
output:
  html_document
---

```{r setup, include=FALSE}
fluidigm_file <- params$fluidigm_file
somalier_file <- params$somalier_file
json_file <- params$json_file

knitr::opts_chunk$set(echo = FALSE, message=FALSE,warning=FALSE,results="asis")
```

```{r lib, echo=FALSE}
suppressPackageStartupMessages({
    library(dplyr)
    library(tidyr)
    library(stringr)
    library(ggplot2)
    library(ggrepel)
    library(gt)
    library(jsonlite)
})
```

```{r read_in_dataframes, echo=FALSE}
# Read in Fluidigm csv generated by lab
fluidigm_df <- read.csv(fluidigm_file, skip = 15, header = TRUE)
plate_barcode <- read.csv(fluidigm_file, header = FALSE)[1,3]
somalier_data <- read.delim(somalier_file, header = TRUE, sep = "\t")

# Y and X chromosome markers
y_markers <- c("rs2032624", "rs1865680", "rs17307398")
x_markers <- c("rs525869", "rs530501", "rs2040962")
```

---
subtitle: "Plate barcode: `r plate_barcode`"
---

```{r zygosity function, echo=FALSE}
# create a function to check zygosity
zygosity <- function(x) {
  l <- length(unique(base::strsplit(x, "")[[1]]))
  if (l == 2) {
    return("homozygous")
             
    } 
  else if (l == 3) {
    return("heterozygous")
              
    }
  else {
    return("No Call")
              
    }
            
  }

```

```{r snp_calls, echo=FALSE}
# select relevant columns and convert calls below 75% to No Call
snp_calls <- fluidigm_df %>% 
  select(Assay,Allele.X,Allele.Y,Name,Confidence,Final,Converted) %>%
  filter(!grepl("NTC", Converted)) %>%
  mutate(Converted = case_when(Confidence < 75 ~ "No Call", .default = Converted))
```

```{r summ_qc, echo=FALSE}
# get non-Y chromosome mean quality and number of missing calls outside of chrY per sample
summ_qc <- snp_calls %>% 
  filter(!Assay %in% y_markers) %>%
  group_by(Name) %>% 
  summarise(mean_confidence = mean(Confidence), 
            markers_called = sum(Converted != "No Call"))

# mark samples with fewer than 75 non-Y markers called or mean confidence below 75 as not passing
not_passed <- summ_qc$Name[summ_qc$markers_called < 75 | summ_qc$mean_confidence < 75]

if (length(not_passed) == 0) {
  not_passed_df <- data.frame("Name" = NA, "Mean confidence" = NA, " Markers called" = NA)
} else {
  not_passed_df <- summ_qc %>%
    filter(markers_called < 75 | mean_confidence < 75) %>%
    dplyr::rename(`Mean confidence` = mean_confidence) %>% dplyr::rename(`Markers called` = markers_called)
}
```

```{r pivot, echo=FALSE}
# remove non-passing samples and pivot df
snp_calls_pivot <- snp_calls %>%
  select(Assay,Name,Converted) %>%
  filter(!Name %in% not_passed) %>%
  pivot_wider(names_from = Assay, values_from = Converted) %>%
  mutate(across(everything(), ~ str_replace(string = .x, 
                                            pattern = ":",
                                            replacement = " ")))
```

```{r sex_calls, echo=FALSE}
# call sex based on zygosity of sex chromosome markers
sex_calls <- snp_calls_pivot %>%
  select(all_of(c("Name",y_markers, x_markers))) %>%
  mutate(across(.cols = starts_with("rs"), ~ sapply(.x, zygosity))) %>%
  mutate(sex = case_when(rs2032624 == "No Call" &  
                         rs1865680 == "No Call" & 
                         rs17307398 == "No Call" ~ "female",
                         (rs2032624 == "homozygous" | 
                          rs1865680 == "homozygous" | 
                          rs17307398 == "homozygous" ) &
                         (rs525869 != "heterozygous" &
                          rs530501 != "heterozygous" &
                          rs2040962 != "heterozygous") ~ "male",
                         .default = "inconclusive"))
    
sex_calls_min <- sex_calls %>% select(Name, sex)

if (dim(sex_calls %>% filter(sex == "inconclusive"))[1] == 0) {
  df_inconclusive <- data.frame("Name" = NA, "rs2032624" = NA, "rs1865680" = NA, "rs17307398" = NA,
                                "rs525869" = NA, "rs530501" = NA, "rs2040962" = NA, "Sex" = NA)
} else {
    df_inconclusive <- sex_calls %>% 
      filter(sex == "inconclusive") %>%
      dplyr::rename(Sex = sex)
}  
```

```{r populate_json, echo=FALSE}
# populate json with sex calls and whether sample passed
report_json <- fromJSON(json_file)

for (s in not_passed) {
      report_json[["samples"]][[s]][["passed"]] <- FALSE
  
}

for (s in sex_calls$Name) {
      report_json[["samples"]][[s]][["passed"]] <- TRUE
  sex <- sex_calls_min[sex_calls_min$Name == s,]$sex
    report_json[["samples"]][[s]][["fluidigm_predicted_sex"]] <- sex
  
}

export_json <- toJSON(report_json, pretty = TRUE, null = 'null', auto_unbox = TRUE)
writeLines(export_json, json_file)
```

```{r matches, echo=FALSE}
matches <- somalier_data %>% 
  filter(relatedness >= 0.8) %>%
  filter(str_split_i(`X.sample_a`, "_", 2) %in% snp_calls$Name | str_split_i(`sample_b`, "_", 2) %in% snp_calls$Name) %>%
  select(`X.sample_a`, sample_b, relatedness, n) %>%
  dplyr::rename(sample_a = `X.sample_a`) %>% dplyr::rename(overlapping_sites = n)

if (dim(matches)[1] == 0) {
    df_matches <- data.frame("Sample A" = NA, "Sample B" = NA, "Relatedness" = NA, "Overlapping sites" = NA)
} else {
    df_matches <- matches %>%
      dplyr::rename(`Sample A` = sample_a) %>% dplyr::rename(`Sample B` = sample_b) %>% dplyr::rename(Relatedness = relatedness, `Overlapping sites` = overlapping_sites)
}
```

### Fluidigm results QC
Samples are considered passing if at least 75 out of 93 markers (excluding chrY markers) are called (horizontal gray line) and if these calls have a mean confidence score of 75 or higher (vertical gray line). Failing samples are labeled and shown in red.
```{r qc_plot, echo=FALSE, fig.height=3.5}
ggplot(summ_qc, aes(x = mean_confidence, y = markers_called)) + 
  theme_bw() +
  geom_point(aes(colour = markers_called >= 75 & mean_confidence >= 75), show.legend = TRUE) +
  scale_colour_manual(name = "Passing", 
  limits = c("TRUE", "FALSE"), 
  values = c("FALSE" = "#F8766D", "TRUE" = "#00BFC4"), drop = FALSE) +
  geom_vline(xintercept = 75, linetype = "dashed", colour = "gray") +
  geom_hline(yintercept = 75, linetype = "dashed", colour = "gray") +
  ylim(0,93) +
  xlim(0,100) +
  ylab("Number of markers called (outside of chrY)") +
  xlab("Mean confidence score (outside of chrY)") +
  geom_text_repel(data = summ_qc %>% filter(Name %in% not_passed),
                  aes(label = Name), color = "red") +
  theme(axis.title = element_text(size = 10))
```

```{r not_passed_table, echo=FALSE}
not_passed_df %>%
  gt() %>%
  tab_header(title = md("**Failed samples**")) %>%
  tab_options(table.font.size = 14, 
              heading.align = "left",
              table.align = "left",
              table.border.top.color = "white", 
              heading.border.bottom.color = "white") %>%
  tab_footnote(footnote = "Includes samples with a mean confidence score below 75 or\nfewer than 75 out of 93 non-Y chromosome markers called.")
```

##### Note: no sex calling or genotype matching is performed for failed samples!

### Fluidigm sex prediction
Sample sex is predicted based on calls for three chrY and three chrX markers. Female samples are expected to have "No Call" for all chrY markers, while male samples should have only homozygous calls for all chrX markers. Sex cannot be determined for samples that do not match this pattern.

```{r inconclusive_table, echo=FALSE}
df_inconclusive %>%
  gt() %>%
  tab_spanner(
    label = "Y markers",
    columns = y_markers
    ) %>%
  tab_spanner(
    label = "X markers",
    columns = x_markers
    ) %>%
  tab_header(title = md("**Samples with inconclusive sex calls**")) %>%
  tab_options(table.font.size = 14, 
              heading.align = "left",
              table.align = "left",
              table.border.top.color = "white", 
              heading.border.bottom.color = "white") %>%
  tab_footnote(footnote = "Sex is inconclusive when a sample has calls on chrY and also has heterozygous sites on chrX.")
```

### Sample relatedness
Sample mixups and matches to previously sequenced samples are identified with Somalier (https://github.com/brentp/somalier). 
Only matches involving at least one sample from the current plate are included in this table.
```{r matches_table, echo=FALSE}
df_matches %>% 
  gt() %>%
  tab_header(title = md("**Samples with genotype matches**"), subtitle =  paste("Current plate: ", plate_barcode)) %>%
  tab_options(table.font.size = 14, 
              heading.align = "left",
              table.align = "left",
              table.border.top.color = "white", 
              heading.border.bottom.color = "white") %>%
  tab_footnote(footnote = "Samples with relatedness of 0.8 or higher are shown.") %>%
  tab_footnote(footnote = "Sample prefix indicates the barcode of the SNP Array on which the sample was run.")
```

