##############################################################################
### Script to convert Fluidigm genotyping output from McGill Genome Centre ###
### Filter by quality, call sex, and output into VCF format.               ###
### Mareike Janiak, Canadian Centre for Computational Genomics             ###
### Last updated 2025-06-17                                                ###
##############################################################################
### Usage : Rscript fluidigm2vcf.R -i input_file -o output_dir

usage <- function(errM) {
    cat("\nUsage : Rscript fluidigm2vcf.R [option]\n")
    cat("       -i      : fluidigm file\n")
    cat("       -o      : output directory\n")
    cat("       -h      : this help\n\n")
    stop(errM)                
}


args <- commandArgs(trailingOnly = T)

## default arg values
fluidigm_file <- ""
out_path <- ""

## get arg variables
for (i in 1:length(args)) {
    if (args[i] == "-i") {
                fluidigm_file <- as.character(args[i+1])

    } else if (args[i] == "-o") {
                out_path <- as.character(args[i+1])
    
    } else if (args[i] == "-h") {
                usage("")
        
    }
    
}

## check that files exist
if (!(file.exists(fluidigm_file))) {
        usage("Error : Fluidigm file not found")

}
if (out_path == "") {
        usage("Error : Output directory not specified")

}

# remove trailing "/" if necessary
tmpOP <- strsplit(out_path, "")
if (tmpOP[[1]][length(tmpOP[[1]])] == "/") {
    out_path <- paste(tmpOP[[1]][1:(length(tmpOP[[1]])-1)], collapse="")

}

suppressPackageStartupMessages({
    library(dplyr)
    library(tidyr)
    library(stringr)
    library(Biostrings)
})

# Read in Fluidigm csv generated by lab
fluidigm_df <- read.csv(fluidigm_file, skip = 15, header = TRUE)
plate_barcode <- read.csv(fluidigm_file, header = FALSE)[1,3]

# Y and X chromosome markers
y_markers <- c("rs2032624", "rs1865680", "rs17307398")
x_markers <- c("rs525869", "rs530501", "rs2040962")

# select relevant columns and convert calls below 75% to No Call
snp_calls <- fluidigm_df %>% 
  select(Assay,Allele.X,Allele.Y,Name,Confidence,Final,Converted) %>%
  filter(!grepl("NTC", Converted)) %>%
  mutate(Converted = case_when(Confidence < 75 ~ "No Call", .default = Converted))
    
# get non-Y chromosome mean quality and number of missing calls outside of chrY per sample
summ_qc <- snp_calls %>% 
  filter(!Assay %in% y_markers) %>%
  group_by(Name) %>% 
  summarise(mean_confidence = mean(Confidence), 
            markers_called = sum(Converted != "No Call"))

# mark samples with fewer than 75 non-Y markers called or mean confidence below 75 as not passing
not_passed <- summ_qc$Name[summ_qc$markers_called < 75 | summ_qc$mean_confidence < 75]

## SNP strand information previously obtained externally from dbsnp bed files
SCRIPT_HOME <- Sys.getenv("SCRIPT_HOME")
snp_strands_file <- paste(SCRIPT_HOME, "/resources/fluidigm_array_snp_loci_strands.tsv", sep = "")
SNPs_df <- read.delim(snp_strands_file, header = TRUE, sep = "\t")
strand_df <- SNPs_df %>% select(Assay, strand)

## Add strand information to snp df
snp_calls <- left_join(snp_calls, strand_df, by="Assay")

snp_calls <- snp_calls %>% 
  mutate(
    Allele.X = case_when(strand == "-" ~ Allele.X %>% DNAStringSet %>% reverseComplement %>% as.character, strand == "+" ~ Allele.X),
    Allele.Y = case_when(strand == "-" ~ Allele.Y %>% DNAStringSet %>% reverseComplement %>% as.character, strand == "+" ~ Allele.Y)
    ) %>%
  mutate(Converted = case_when(Converted == "No Call" ~ "N:N", .default = Converted)) %>%
  separate(., Converted, into = c("Allele1", "Allele2"), sep = ":", remove = TRUE) %>%
  mutate(
    Allele1 = case_when(strand == "-" ~ Allele1 %>% DNAStringSet %>% reverseComplement %>% as.character, strand == "+" ~ Allele1),
    Allele2 = case_when(strand == "-" ~ Allele2 %>% DNAStringSet %>% reverseComplement %>% as.character, strand == "+" ~ Allele2)
    ) %>%
  unite("Converted", Allele1:Allele2, sep = ":") %>%
  mutate(Converted = str_replace(string = Converted, pattern = "N:N",
  replacement = "No Call"))
          
# Combine genomic position info with lab df and convert calls to be consistent with what is the ref and alt allele
df_all <- left_join(snp_calls, SNPs_df, by="Assay") %>%
  filter(!Name %in% not_passed) %>%
  mutate(Converted = case_when(Converted == "No Call" ~ "N:N", .default = Converted)) %>%
  separate(., Converted, into = c("Allele1", "Allele2"), sep = ":", remove = TRUE) %>%
  mutate(ALT = case_when(Allele.X == REF ~ Allele.Y, Allele.X != REF ~ Allele.X)) %>%
  mutate(Call1 = case_when(Allele1 == "N" ~ ".", Allele1 == REF ~ "0", Allele1 != REF ~ "1")) %>%
  mutate(Call2 = case_when(Allele2 == "N" ~ ".", Allele2 == REF ~ "0", Allele2 != REF ~ "1")) %>%
  unite("Call", Call1:Call2, sep = "/") %>%
  mutate(Call = case_when(Call == "1/0" ~ "0/1", Call != "1/0" ~ Call)) %>%
  dplyr::select(Assay, Name, chr_name, chrom_start, REF, ALT, Call) %>%
  mutate(chrom_start = as.character(chrom_start)) %>%
  mutate(Call = paste0(Call,":20")) %>%
  pivot_wider(names_from = Name, values_from = Call) %>%
  dplyr::rename(`#CHROM` = chr_name) %>% dplyr::rename(POS = chrom_start) %>% dplyr::rename(ID = Assay)

# create VCF format to be combined with header
vcf <- df_all %>%
  arrange(`#CHROM`) %>%
  mutate(`#CHROM` = paste0("chr",`#CHROM`)) %>%
  relocate(ID, .after = POS) %>%
  mutate(QUAL = ".", .after = ALT) %>%
  mutate(FILTER = "PASS", .after = QUAL) %>%
  mutate(INFO = paste0("RSID=",ID), .after = FILTER) %>%
  mutate(FORMAT = "GT:DP", .after = INFO)

# create VCF by appending vcf df to header
write.table(vcf, file = paste(out_path,"/fluidigm.", plate_barcode,".vcf",sep=""), quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE, append = TRUE)


